### DISCUSSION TITLE
"Nothing to do" during interactive rebase

### DISCUSSION THREAD

**Discussion by chrhansk** [magit/magit#4969] [Support] [answered]
### Situation

I want to use an interactive rebase to remove the single most recent (or the x most recent) commit(s) from history. To this end I

1. Bring up `magit-status` using `C-x g` in some project file
2. Navigate to the `Recent commits` section
3. Move the cursor to the most recent commit
4. Hit `r` `i` to start the interactive rebasing
5. Move the cursor to the initial line containing the (single) commit hash / message
6. Press `k` to mark the commit to be dropped
7. Press `C-c C-c` to tell Git to make it happen

### Expected behavior

The commit is removed from history, and I am thrown back to the `magit-status` window

### Actual behavior

I an thrown back to the `magit-status` with a `GitError` showing `nothing to do  [Type ``$' for details]`. The `magit-process`, brough up with `$`, shows

```
hint: Waiting for your editor to close the file...
Waiting for Emacs...
error: nothing to do
```

The commit is not removed from the history (apparently, the history does not change at all). 

### Backtrace

I entered `toggle-debug-on-error` and executed the same sequence of commands which did not produce an error message. This seems to be an error in a sub command launched by git (?). I suspect that this issue is some corner case in the interactive rebase implementation.

### Output of `magit-version`

```
Magit 20230723.1601 [>= 3.3.0.50-git], Transient 0.3.7.50-git, Git 2.41.0, Emacs 28.2, gnu/linux
```

### Plain Emacs

This defect is reproducible on my system using `emacs -Q`

**Comment by tarsius** [comment, marked as answer]
> 1. Bring up `magit-status` using `C-x g` in some project file
> 2. Navigate to the `Recent commits` section
> 3. Move the cursor to the most recent commit
> 4. Hit `r` `i` to start the interactive rebasing
> 5. Move the cursor to the initial line containing the (single) commit hash / message
> 6. Press `k` to mark the commit to be dropped
> 7. Press `C-c C-c` to tell Git to make it happen

Do this instead:

1. Bring up `magit-status` using `C-x g` in some project file
2. Navigate to the `Recent commits` section
3. Move the cursor to the **second** most recent commit
4. Hit `X` `h` to reset to **that** commit in one go

> ```
> hint: Waiting for your editor to close the file...
> Waiting for Emacs...
> error: nothing to do
> ```
> 
> The commit is not removed from the history (apparently, the history does not change at all).

Nothing about the commit, **that are rewritten**, changes. You forgo applying the very last commit, but when Git checks whether "anything changed", it does not look at that commit, just the ones you told it to apply and since you are instructing Git to replay them exactly as before, it can actually just fast-forward to the latest of these commits, i.e., there is "nothing to do".

But this behavior is indeed a bit counterintuitive. For example you might want to sign the last n commits, without rearranging them or anything like that. In such cases you need to use `--force-rebase` to tell Git "Well actually, this makes a difference. Try and see.".

**Reply by chrhansk** [reply]
Well, you certainly have a point there, and maybe this is a rough edge in git itself... But it is a bit unfortunate that this command (which should, in principle be valid) causes an error message to appear.
